gix-traverse/src/commit/topo/mod.rs

```diff
@@ -0,0 +1,79 @@
+//! Topological commit traversal, similar to `git log --topo-order`.
+
+use gix_hash::ObjectId;
+use gix_revwalk::{graph::IdMap, PriorityQueue};
+
+use bitflags::bitflags;
+
+use super::Parents;
+
+/// A commit walker that walks in topographical order, like `git rev-list
+/// --topo-order` or `--date-order` depending on the chosen [`Sorting`].
+pub struct Walk<Find, Predicate> {
+    commit_graph: Option<gix_commitgraph::Graph>,
+    find: Find,
+    predicate: Predicate,
+    indegrees: IdMap<i32>,
+    states: IdMap<WalkFlags>,
+    explore_queue: PriorityQueue<iter::GenAndCommitTime, ObjectId>,
+    indegree_queue: PriorityQueue<iter::GenAndCommitTime, ObjectId>,
+    topo_queue: iter::Queue,
+    parents: Parents,
+    min_gen: u32,
+    buf: Vec<u8>,
+}
+
+/// The errors that can occur during creation and iteration.
+#[derive(thiserror::Error, Debug)]
+#[allow(missing_docs)]
+pub enum Error {
+    #[error("Indegree information is missing")]
+    MissingIndegreeUnexpected,
+    #[error("Internal state (bitflags) not found")]
+    MissingStateUnexpected,
+    #[error(transparent)]
+    CommitGraphFile(#[from] gix_commitgraph::file::commit::Error),
+    #[error(transparent)]
+    ObjectDecode(#[from] gix_object::decode::Error),
+    #[error(transparent)]
+    Find(#[from] gix_object::find::existing_iter::Error),
+}
+
+bitflags! {
+    /// Set of flags to describe the state of a particular commit while iterating.
+    // NOTE: The names correspond to the names of the flags in revision.h
+    #[repr(transparent)]
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+    struct WalkFlags: u32 {
+        /// Commit has been seen
+        const Seen = 0b000001;
+        /// Commit has been processed by the Explore walk
+        const Explored = 0b000010;
+        /// Commit has been processed by the Indegree walk
+        const InDegree = 0b000100;
+        /// Commit is deemed uninteresting for whatever reason
+        const Uninteresting = 0b001000;
+        /// Commit marks the end of a walk, like `foo` in `git rev-list foo..bar`
+        const Bottom = 0b010000;
+        /// Parents have been processed
+        const Added = 0b100000;
+    }
+}
+
+/// Sorting to use for the topological walk
+#[derive(Clone, Copy, Debug, Default)]
+pub enum Sorting {
+    /// Show no parents before all of its children are shown, but otherwise show
+    /// commits in the commit timestamp order.
+    #[default]
```

Comment:
A quick question: Since this is all about `topo`, should the default not be `TopoOrder`?

I also have trouble understanding how this `DateOrder` maps to Git. Is there a `git rev-list` equivalent?

Reply:
I don't have Git's code (with my debug modifications) in front of me. I can fill in details later, but from memory:

In Git there is an older sorting algorithm that that's used if gen numbers aren't available. If neither `--date-order` nor `--topo-order` is specified then _that_ algorithm is used, for some reason. So in Git the default is "neither of them". (this is the part I must double-check). I think the idea is that with time the old algorithm can be removed from Git entirely, but it hasn't happened yet. 

`DateOrder` maps to `git rev-list --date-order`. I have no strong feelings about which is the default for us; I imagine most callers would want to set the order explictly, anyway. Git's default is equivalent to the new algorithm with `--date-order` (again, I need to double check this).

Reply:
Wonderful, thanks so much for the explanation! The level of detail is perfectly fine, and I just added b8486f7c to add another test and incorporate the information.

It's amazing that Git still has this older algorithm, but then added more just by means of flags which don't seem to be mutually exclusive even - ` git rev-list --topo-order --date-order @` does *something*.